Fixed an issue of data that contain tp tag sometimes recognized as flow

ilyasoifer opened the pull request:
Fixes #8335

ilyasoifer commented:
Note to self - still need to validate that minimap2 data does not get recognized as flowBased

droazen commented:
Thanks @ilyasoifer ! Was the test bam you added aligned with minimap2? If not, we should make sure to add at least one simple HaplotypeCaller regression test that takes an actual minimap2-aligned input.

ilyasoifer commented:
@droazen the bam that I added was aligned with minimap2. However, it is a really small BAM, so I could not run it through the integration tests. It sounds useful to add minimap2-aligned bam to HaplotypeCallerIntegrationTest. Can you or @jamesemery point me to a suitable BAM for that?

jamesemery commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:
```diff
@@ -135,16 +134,23 @@ public static void finalizeRegion(final AssemblyRegion region,
         for (final GATKRead originalRead : region.getReads()) {
+
+            boolean isFlowBasedData = false;
+            if (FlowBasedReadUtils.hasFlowTags(originalRead)) {
```
Can this be condensed into a single method so external tools don't have to know about the readgroup info caching? Can you make a method `FlowBasedReadUtils.isThisFlowBased()` that extracts this entire operation away and handles your readgroup caching internally?

jamesemery reviewed (changes requested):
@ilyasoifer Thank you for doing this. A few more defensive checks and future-proofing to hopefully save us from having this headache again in the future. Could you realign our HaplotypeCaller input bam snippets with minimap 2 and make a quick integration test asserting that they still funciton?

ilyasoifer commented on src/main/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtils.java:
```diff
@@ -26,17 +28,26 @@ public class FlowBasedReadUtils {
+            if (readGroup.getPlatform()==null){
+                isFlowPlatform = false;
+            } else {
+                isFlowPlatform = NGSPlatform.valueOf(readGroup.getPlatform()).getSequencerType() == SequencerFlowClass.FLOW;
```
Old Ultima data used PL:LS454 because ULTIMA was not a legal platform. I think I prefer allowing LS454 and ULTIMA to work this way given low chance anyone is now running LS454 data

jamesemery replied:
Can you special case the PL:LS454 platform and give an explanation about it here? Again, I want to be as specific as possible about the inputs here since this is a greedy non opt-in behavior we should be careful we don't start clobbering somebodies inputs in all cases. If it IS labeled as 454 data can you check for the other fields that should make it readable as ultima data?

ilyasoifer replied:
Added tests

ilyasoifer commented on src/main/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtils.java:
```diff
@@ -26,17 +28,26 @@ public class FlowBasedReadUtils {
+                isFlowPlatform = NGSPlatform.valueOf(readGroup.getPlatform()).getSequencerType() == SequencerFlowClass.FLOW;
+            }
+            if (isFlowPlatform) {
+                Utils.nonNull(readGroup);
```
I think I fixed this, can you check?

ilyasoifer commented on src/test/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtilsUnitTest.java:
```diff
@@ -24,9 +21,33 @@ void testReadGroupParsing() throws Exception{
+        SAMReadGroupRecord rg1 = header.getReadGroup("test");
+        SAMRecord read = reader.iterator().next();
+        GATKRead gread = new SAMRecordToGATKReadAdapter(read);
+        assert(FlowBasedReadUtils.hasFlowTags(gread));
```
Fixed

ilyasoifer commented:
@droazen , @jamesemery  - I think I addressed your comments, back to you now. 
Looking at this deeper, I think we will anyway remove the "removeUncertainFlows" options that triggered all this
It resulted in a need to fix a bug we had in a basecalling, but it had been long resolved. 

In any case, I think that the check is more robust now, and it was a good fix.

ilyasoifer commented:
> @ilyasoifer do you have the means to quickly run minimap2? I would recommend simply realigning src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam with minimap2 and making a quick "are tests consistent with pervious versions" test with a checked in vcf output. I don't know how to wrangle minimap2 to handle mates correctly however so i don't know if this is easy

Added test like this

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtils.java:
```diff
@@ -26,17 +29,36 @@ public class FlowBasedReadUtils {
+                // we require also FO field to consider ReadGroup to come from the flow
+                boolean tmp = (NGSPlatform.fromReadGroupPL(readGroup.getPlatform()) == NGSPlatform.LS454) ||
+                        (NGSPlatform.fromReadGroupPL(readGroup.getPlatform()) == NGSPlatform.ULTIMA);
+                tmp = tmp & (readGroup!=null);
```
readGroup!=null here after referencing it several times above?

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtils.java:
```diff
@@ -118,22 +141,28 @@ public static int flowSumOfBaseQualities(final GATKRead read) {
 
     }
 
+    public static boolean isFlowPlatform(final SAMFileHeader hdr, final GATKRead read) {
```
Can you just add a javadoc to this?

jamesemery replied:
And also possibly update the class javadoc to point people to this method as the definitive "is this flow based and should i trigger that code" method

ilyasoifer replied:
fixed

jamesemery commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java:
```diff
@@ -101,7 +101,37 @@ public void testVCFModeIsConsistentWithPastResults(final String inputFileName, f
+    /*
+     * Test that minimap2 data are supported and consistent with past results
+     */
+    public void testVCFModeMinimap2IsConsistentWithPastResults() throws Exception {
```
this doesn't get run without adding the @test tag

ilyasoifer replied:
Fixed

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtils.java:
```diff
@@ -26,17 +29,36 @@ public class FlowBasedReadUtils {
+                isFlowPlatform = tmp;
+
+                //in addition validate that the ultima data has FO tag, otherwise break
+                if ((NGSPlatform.fromReadGroupPL(readGroup.getPlatform()) == NGSPlatform.ULTIMA) & (!isFlowPlatform)){
```
the logic here is a little confusing though I believe it accomplishes what you want. You might consider refactoring this a little so its cleaner but its not critical.

ilyasoifer replied:
Fixed

jamesemery reviewed (changes requested):
A few minor comments and one test update and then i think it looks good. Thank you for the quick turnaround on fixing this!

jamesemery reviewed (changes requested):
A few minor comments and one test update and then i think it looks good. Thank you for the quick turnaround on fixing this!

ilyasoifer commented on src/test/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtilsUnitTest.java:
```diff
@@ -24,9 +26,117 @@ void testReadGroupParsing() throws Exception{
+        tests.add(new Object[]{hdr, read3, false});
+
+        GATKRead read4 = makeRead(new byte[]{'T','A','G','C','G','A'}, false);
+        read4.setAttribute("tp",new byte[6]);
```
Done

ilyasoifer commented:
@jamesemery I believe I addressed your comments, PTAL

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/read/FlowBasedReadUtils.java:
```diff
@@ -26,17 +36,35 @@ public class FlowBasedReadUtils {
+                isFlowPlatform = false;
+            } else if (NGSPlatform.fromReadGroupPL(readGroup.getPlatform())==NGSPlatform.UNKNOWN){
+                isFlowPlatform = false;
+            } else if (NGSPlatform.fromReadGroupPL(readGroup.getPlatform()) == NGSPlatform.LS454) {
```
I know 454 data is ancient at this point but i would still prefer this LS454 case to explicitly check for 
```
if (readGroup.getFlowOrder()!=null) {
                    isFlowPlatform = true; else false;
```
I'm not sure i trust the Ultima flow ordering code is going to handle 454 data well without testing it

ilyasoifer replied:
I think this is not this check is supposed to do. Old Ultima data were labeled LS454 and initially did not have the FO tag. Afterwards the tag was added .
The full check is done by isFlowPlatform function that also verifies that the reads have tp tag.

jamesemery replied:
Ahh i see... horrifying... well i'm alright to let this through in this case then

jamesemery reviewed (commented):
One FINAL gripe and then its good to go. Thank you for going back and forth on this.

jamesemery reviewed (approved):
Alright this should be good 👍

ilyasoifer commented:
Thanks for the help!

AishaShah commented:
Is this fixed in version 4.4.0.0 of gatk?

rmormando commented:
Hi! I know this issue is closed and a solution has been found but I am facing this same issue right now and wanted to follow up and see if it's been patched in the current 4.4.0.0 version of gatk? Or is the new release expected to be published soon?
